pageserver/src/layered_repository.rs

```diff
@@ -0,0 +1,1145 @@
+        };
+        Self::save_metadata(self.conf, timelineid, self.tenantid, &metadata)?;
+
+        let timeline = LayeredTimeline::new(
+            self.conf,
+            metadata,
+            None,
+            timelineid,
+            self.tenantid,
+            self.walredo_mgr.clone(),
+        )?;
+
+        let timeline_rc = Arc::new(timeline);
+        let r = timelines.insert(timelineid, timeline_rc.clone());
+        assert!(r.is_none());
+        Ok(timeline_rc)
+    }
+
+    /// Branch a timeline
+    fn branch_timeline(&self, src: ZTimelineId, dst: ZTimelineId, start_lsn: Lsn) -> Result<()> {
+        let src_timeline = self.get_timeline(src)?;
+
+        // Create the metadata file, noting the ancestor of the new timeline.
+        // There is initially no data in it, but all the read-calls know to look
+        // into the ancestor.
+        let metadata = TimelineMetadata {
+            last_valid_lsn: start_lsn,
+            last_record_lsn: start_lsn,
+            prev_record_lsn: src_timeline.get_prev_record_lsn(),
+            ancestor_timeline: Some(src),
+            ancestor_lsn: start_lsn,
+        };
+        std::fs::create_dir_all(self.conf.timeline_path(&dst, &self.tenantid))?;
+        Self::save_metadata(self.conf, dst, self.tenantid, &metadata)?;
+
+        info!("branched timeline {} from {} at {}", dst, src, start_lsn);
+
+        Ok(())
+    }
+
+    //
+    // How garbage collection works:
+    //
+    //                    +--bar------------->
+    //                   /
+    //             +----+-----foo---------------->
+    //            /
+    // ----main--+-------------------------->
+    //                \
+    //                 +-----baz-------->
+    //
+    //
+    // 1. Grab a mutex to prevent new timelines from being created
+    // 2. Scan all timelines, and on each timeline, make note of the
+    //    all the points where other timelines have been branched off.
+    //    We will refrain from removing page versions at those LSNs.
+    // 3. For each timeline, scan all snapshot files on the timeline.
+    //    Remove all files for which a newer file exists and which
+    //    don't cover any branch point LSNs.
+    //
+    // TODO:
+    // - if a relation has been modified on a child branch, then we
+    //   don't need to keep that in the parent anymore. But currently
+    //   we do.
+    //
+    // - Currently, this is only triggered manually by the 'do_gc' command.
+    //   There is no background thread to do it automatically.
+    fn gc_iteration(
+        &self,
+        target_timelineid: Option<ZTimelineId>,
+        horizon: u64,
+        _compact: bool,
+    ) -> Result<GcResult> {
+        let mut totals: GcResult = Default::default();
+        let now = Instant::now();
+
+        // grab mutex to prevent new timelines from being created here.
+        // TODO: We will hold it for a long time
+        let mut timelines = self.timelines.lock().unwrap();
```

Comment:
Perhaps. I'm not sure if there would be race conditions if a new branch was created.

One quick thing we could do is to use an RwLock instead of Mutex. That would already help a lot, because then the GC would only block creating new branches but existing branches could be accessed concurrently.